### ISSUE TITLE
When creating an order in the backend, the customer group always defaults to the default customer group even when overridden

### ISSUE THREAD

**Issue by justinharris1986** [magento/magento2#22121]
### Preconditions
1. Magento 2.2.8
2. PHP 7.1.27 (In our case - may or may not be a factor - we are running in php-fpm)
3. MariaDB 10.3.13-1.el7.centos (In our case - may or may not be a factor)
4. at least 2 customer groups
5. (Optional, but illustrates the issue better) A customer not in the "Default group" Located in Stores -> Configuration ->Customers -> Customer Configuration -> Create New Account Options -> Default Group. 

### Steps to reproduce
1. Create a 2nd customer group
2. Create a customer, place in this 2nd group, created above
3. Go to Sales -> Orders
4. Create New order
5. Select your customer
6. Observe that the customer group is the default one configured under Stores -> Configuration ->Customers -> Customer Configuration -> Create New Account Options -> Default Group. 
![image](https://user-images.githubusercontent.com/7268384/55429996-e5a93500-555a-11e9-8edb-90ee170254a5.png)
7. Attempt to change this from the default to the 2nd customer group created in step 1. Page refreshes to default customer group.

### Expected result
1.  The customer group defaults to the actual customer group configured under "Steps to Reproduce, step 2" This was how it operated in 2.2.7
2. When overriding the customer group dropdown, it should stick, not refresh back to the default group

### Actual result
1. the customer group defaults to the "default group" as configured under Stores -> Configuration ->Customers -> Customer Configuration -> Create New Account Options -> Default Group.
2. when you manually change the customer group dropdown in "Steps to Reproduce, step 7" the group sticks with the requested group and all pricing updates to the customer group

**Comment by justinharris1986** [comment]
@magento-engcom-team give me 2.2-develop instance

**Comment by magento-engcom-team** [comment]
Hi @justinharris1986. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @justinharris1986, here is your [Magento instance](https://i-22121-2-2-develop.instances.magento-community.engineering).
Admin access: [https://i-22121-2-2-develop.instances.magento-community.engineering/admin](https://i-22121-2-2-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by justinharris1986** [comment]
@magento-engcom-team  Yes I was able to reproduce it.

I forgot that magento had other customer groups, you *CANNOT* use any of the built on ones. You must create a new one. 

When creating them in my store, I see the other customer groups as dropdowns. When creating them in your environment, I do not see them in the dropdowns.
This is when recreating in yours:
![image](https://user-images.githubusercontent.com/7268384/55430594-84826100-555c-11e9-9876-c9767aa3fe74.png)

However,take a look in the back end I created this new group you do not see above:
![image](https://user-images.githubusercontent.com/7268384/55430624-982dc780-555c-11e9-9d66-6032187aef10.png)

**Comment by justinharris1986** [comment]
I have found if you delete the default groups (ie delete wholesale) you still see it on the order screen, and when you try to select it, it errors out: 
![image](https://user-images.githubusercontent.com/7268384/55431042-80a30e80-555d-11e9-9152-cc98d5e51584.png)

![image](https://user-images.githubusercontent.com/7268384/55431050-8993e000-555d-11e9-9067-ebef1d8f036a.png)

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-backlog-nazar 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-99044`, `MAGETWO-99045` were created

**Issue Available**: @engcom-backlog-nazar, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by nickpiro** [comment]
I can confirm this issue is also happening on 2.2.8

**Comment by mystix** [comment]
Managed to trace this bug to the private method `extractValuesFromAttributes()` introduced in M2.2.8.

File:
`https://github.com/magento/magento2/blob/2.2.8/app/code/Magento/Sales/Block/Adminhtml/Order/Create/Form/Account.php`:
```
    /**
     * Extract the form values from attributes.
     *
     * @param array $attributes
     * @param int $storeId
     * @return array
     */
    private function extractValuesFromAttributes(array $attributes, int $storeId): array
    {
        $formValues = $this->getFormValues();
        foreach ($attributes as $code => $attribute) {
            $defaultValue = $attribute->getDefaultValue();
            if (isset($defaultValue) && !isset($formValues[$code])) {
                $formValues[$code] = $defaultValue;
            }
            if ($code === 'group_id' && empty($defaultValue)) { // herein lies the problem
                $formValues[$code] = $this->getDefaultCustomerGroup($storeId);
            }
        }
        return $formValues;
    }
```

The check for an empty `$defaultValue` for the `group_id` attribute is problematic because it fails to also check if `$formValues['group_id']` is empty before overriding it with the system-configured default customer group. 🤦‍♂️ 

The correct implementation for the method `extractValuesFromAttributes()` should thus be:
```
    /**
     * Extract the form values from attributes.
     *
     * @param array $attributes
     * @param int $storeId
     * @return array
     */
    private function extractValuesFromAttributes(array $attributes, int $storeId): array
    {
        $formValues = $this->getFormValues();

        foreach ($attributes as $code => $attribute) {
            if (!isset($formValues[$code])) {
                $defaultValue = $attribute->getDefaultValue();

                if (isset($defaultValue)) {
                    $formValues[$code] = $defaultValue;
                } else if ($code === 'group_id') {
                    $formValues[$code] = $this->getDefaultCustomerGroup($storeId);
                }
            }
        }

        return $formValues;
    }
```

**Comment by nickpiro** [comment]
@mystix this fixed the issue for me. Going to test some orders but your reasoning and change makes sense! Someone needs to add this as a fix. @engcom-backlog-nazar

**Comment by mystix** [comment]
Interestingly, in M2.3.1, this code was "fixed" so the system-configured default customer group is never used 🤦‍♂️🤦‍♂️:

File:
`https://github.com/magento/magento2/blob/2.3.1/app/code/Magento/Sales/Block/Adminhtml/Order/Create/Form/Account.php`

Diff:
https://github.com/magento/magento2/commit/a2688893e7a690ded6524b4ef382fd729d712236#diff-810242e5c9175d4f3571f34d6d47a6eb

Both M2.2 and M2.3 branches should support using the system-configured default customer group.

**Comment by nickpiro** [comment]
@mystix I will never understand the update process of M2 🤦‍♂️

But thank you for helping fix 2.2.8 for me! 

It seems instead of doing the fix you did, they just removed it completely.

If you're interested, I have another strange issue if you want to look into it! :)
https://github.com/magento/magento2/issues/21985

**Comment by mallikaverma** [comment]
I am working on this at #dmcdindia1

**Comment by engcom-Echo** [comment]
Unfortunately, I can't reproduce an issue. I've got an expected result.
On order page in admin panel, the customer group corresponds to the group to which the customer belongs.
If a customer group was deleted not present in the drop-down.

![Peek 2020-12-30 13-19](https://user-images.githubusercontent.com/51681435/103348470-c7ab9400-4aa2-11eb-8a6e-7ac8ae993f29.gif)

Please feel free to comment, reopen or create new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
if you are still facing this issue on the latest `2.4-develop` branch. Thank you for collaboration.

**Comment by magento-engcom-team** [comment]
Hi @justinharris1986. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-24457` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.
